### PR TITLE
Avoid saving `nextBuildNumber` while loading `Job`

### PR DESCRIPTION
**Pull request by jglick**
I spent a while looking into a semi-reproducible test failure affecting CloudBees CI running in high availability mode. The scenario involves a job being created and then a build immediately being scheduled; `config.xml` is created and then `nextBuildNumber` is created shortly thereafter. If another process is loading the job in between, it can attempt to write out `nextBuildNumber` itself, despite not actually running any build of the job. The details are probably not relevant to OSS Jenkins users but the essential guideline is that you should not be writing files _to_ disk just because you are loading some configuration _from_ disk, at least not without a good reason. The file should be written only if and when a build is scheduled; its absence should be treated as an implicit value of 1. (`lastBuildNumber` with an implicit value of 0 would perhaps have been more intuitive.)

I did keep the clause which recomputes and writes out `nextBuildNumber` in case the file existed but was corrupt (not parsable as an integer), since otherwise you might be printing an error on every Jenkins startup, if the job is never built again.

### Testing done

Seems to resolve a timing-dependent test failure observed in CloudBees CI proprietary code.

### Proposed changelog entries

N/A, likely not interesting.

Before the changes are marked as `ready-for-merge`:

```[tasklist]
### Maintainer checklist
- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).
```

### CODE REVIEW

File: core/src/main/java/hudson/model/Job.java
Comment (jglick) [inline core/src/main/java/hudson/model/Job.java:226]: A comment about Hudson 1.27 did not seem particularly helpful at this point!

File: core/src/main/java/hudson/model/Job.java
Comment (jglick) [inline core/src/main/java/hudson/model/Job.java:235]: It no longer does so, so simplify and use the same logic for all `Job` types. (In practice all jobs in common usage use `LazyBuildMixIn`.)

File: core/src/main/java/hudson/model/Job.java
Comment (jglick) [inline core/src/main/java/hudson/model/Job.java:245]: In fact `doCreateItem` calls `onCreatedFromScratch`, which did _not_ create this file, so that was misleading.

File: core/src/main/java/hudson/model/Job.java
Comment (jglick) [inline core/src/main/java/hudson/model/Job.java:237]: Can happen if the file was absent when a job which had never been built was loaded from disk. The field initializer is not called by XStream.

File: core/src/main/java/jenkins/model/lazy/LazyBuildMixIn.java
Comment (jglick) [inline core/src/main/java/jenkins/model/lazy/LazyBuildMixIn.java:112]: compare #2439

File: core/src/main/java/jenkins/model/lazy/LazyBuildMixIn.java
Comment (jglick) [inline core/src/main/java/jenkins/model/lazy/LazyBuildMixIn.java:113]: introduced in #9019

**Review by Vlatombe — APPROVED** [review]

**Review by timja — APPROVED** [review]
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Pull request merged**
